orttraining/orttraining/include/runner/orttraining_c_api.h

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) SignalPop LLC. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include <string.h>
+#include <runner\training_runner.h>
+
+using namespace onnxruntime;
+using namespace onnxruntime::common;
+using namespace std;
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum OrtTrainingOptimizer {
+    ORT_TRAINING_OPTIMIZER_SGD = 0
+};
+
+typedef enum OrtTrainingLossFunction {
+    ORT_TRAINING_LOSS_FUNCTION_SOFTMAXCROSSENTROPY = 0
+};
+
+typedef enum OrtTrainingStringParameter {
+    ORT_TRAINING_MODEL_PATH = 0,
+    ORT_TRAINING_LOG_PATH = 1,
+    ORT_TRAINING_INPUT_LABELS = 2,
+    ORT_TRAINING_OUTPUT_PREDICTIONS = 3,
+    ORT_TRAINING_OUTPUT_LOSS = 4
+};
+
+typedef enum OrtTrainingLongParameter {
+  ORT_TRAINING_NUM_TRAIN_STEPS = 0,
+  ORT_TRAINING_TRAIN_BATCH_SIZE = 1,
+  ORT_TRAINING_EVAL_BATCH_SIZE = 2,
+  ORT_TRAINING_EVAL_PERIOD = 3,
+  ORT_TRAINING_DISPLAY_LOSS_STEPS = 4
+};
+
+typedef enum OrtTrainingNumericParameter {
+    ORT_TRAINING_LEARNING_RATE = 0,
+};
+
+typedef enum OrtTrainingBooleanParameter {
+  ORT_TRAINING_USE_GIST = 0,
+  ORT_TRAINING_USE_CUDA = 1,
+  ORT_TRAINING_USE_PROFILER = 2,
+  ORT_TRAINING_USE_TENSORBOARD = 3,
+  ORT_TRAINING_IS_PERFTEST = 4,
+  ORT_TRAINING_SHUFFLE_DATA = 5
+};
+
+typedef enum OrtDataUse {
+  ORT_DATAUSE_TRAINING = 0,
+  ORT_DATAUSE_TESTING = 1
+};
+
+struct OrtValueCollection;
+
+typedef void(__stdcall* OrtErrorFunctionCallback)(_In_ const size_t ulCount, _In_ OrtValueCollection* output);  // Array of outputs (e.g. label, predictions, loss).
+typedef void(__stdcall* OrtEvaluationFunctionCallback)(size_t num_samples, size_t step);
+typedef void(__stdcall* OrtDataGetBatchCallback)(_In_ const size_t ulBatchSize, _In_ const size_t ulCount, _In_ OrtValueCollection* data); // Array of input locations, then set by call-back implemenation by calling SetAt.
+
+
+// The actual types defined have an Ort prefix
+//ORT_RUNTIME_CLASS(OrtValueCollection);
+struct OrtValueCollection {
+  std::vector<OrtValue*> m_rgValues;
+  std::vector<std::string> m_rgNames;
+
+  OrtValueCollection(int nCapacity) : m_rgValues(nCapacity), m_rgNames(nCapacity) {}
+  ~OrtValueCollection() {}
+
+  size_t Capacity() {
+    return m_rgValues.capacity();
+  }
+
+  size_t Count() {
+    return m_rgValues.size();
+  }
+
+  std::vector<OrtValue*> ValuePtrs() {
+    return m_rgValues;
+  }
+
+  std::vector<std::string> Names() {
+    return m_rgNames;
+  }
+
+  void BeforeUsingAsInput(int queue_id) {
+    for (size_t i = 0; i < m_rgValues.size(); i++) {
+      if (m_rgValues[i]->Fence())
+        m_rgValues[i]->Fence()->BeforeUsingAsInput(onnxruntime::kCpuExecutionProvider, queue_id);
+    }
+  }
+};
+
+struct OrtTrainingApi;
+typedef struct OrtTrainingApi OrtTrainingApi;
+
+
+// The actual types defined have an Ort prefix
+//ORT_RUNTIME_CLASS(TrainingParameters);
+struct OrtTrainingParameters {
```

Comment:
>OrtTrainingParameters [](start = 7, length = 21)

This is not a POD and does not qualify to be in the C header. We provide C++ wrappers on top of the C API

Reply:
I can remove the constructor and destructor and then do the creation and cleanup in the CC files for initialization and cleanup.  Will that work?